designer/server/src/main/scala/pl/touk/nussknacker/ui/process/deployment/DeploymentServiceImpl.scala

```diff
@@ -72,77 +72,134 @@ class DeploymentServiceImpl(
+            validatedComment,
+            ctx.processDetails,
+            ctx.buildInfoProcessingType
+          ) {
+            dispatcher
+              .deploymentManagerUnsafe(ctx.processDetails.processingType)
+              .processCommand(command)
+          }
+          Future.successful(deploymentFuture)
       }
-    }
+    } yield result
+  }
+
+  /**
+    * Common validations and operations for a command execution.
+    * @return gathered data for further command execution
+    */
+  private def prepareCommandContextWithAction[PS: ScenarioShapeFetchStrategy](
+      processId: ProcessIdWithName,
+      actionType: ProcessActionType,
+      getVersionOnWhichActionIsDone: ScenarioWithDetailsEntity[PS] => Option[VersionId],
+      getBuildInfoProcessingType: ScenarioWithDetailsEntity[PS] => Option[ProcessingType]
+  )(implicit user: LoggedUser, ec: ExecutionContext): Future[CommandContext[PS]] = {
+    implicit val freshnessPolicy: DataFreshnessPolicy = DataFreshnessPolicy.Fresh
+    val actionName                                    = ScenarioActionName(actionType)
+    dbioRunner.runInTransaction(
+      for {
+        // 1.1 TODO common parameter validations
+        // 1.2 lock for critical section
+        _ <- actionRepository.lockActionsTable
+        // 1.3. fetch scenario data
+        processDetailsOpt <- processRepository.fetchLatestProcessDetailsForProcessId[PS](processId.id)
+        processDetails    <- existsOrFail(processDetailsOpt, ProcessNotFoundError(processId.name))
+        // 1.4. calculate which scenario version is affected by the action: latest for deploy, deployed for cancel
+        versionOnWhichActionIsDone = getVersionOnWhichActionIsDone(processDetails)
+        buildInfoProcessingType    = getBuildInfoProcessingType(processDetails)
+        // 1.5. check if action is performed on proper scenario (not fragment, not archived)
+        _ = checkIfCanPerformActionOnScenario(actionName, processDetails)
+        // 1.6. check if action is allowed for current state
+        inProgressActionTypes <- actionRepository.getInProgressActionTypes(processDetails.processId)
+        processState          <- getProcessState(processDetails, inProgressActionTypes)
+        _ = checkIfCanPerformActionInState(actionName, processDetails, processState)
+        // 1.7. create new action, action is started with "in progress" state, the whole command execution can take some time
+        actionId <- actionRepository.addInProgressAction(
+          processDetails.processId,
+          actionType,
+          versionOnWhichActionIsDone,
+          buildInfoProcessingType
+        )
+      } yield CommandContext(processDetails, actionId, versionOnWhichActionIsDone, buildInfoProcessingType)
+    )
   }
 
+  // TODO: Use buildInfo explicitly instead of ProcessingType-that-is-used-to-calculate-buildInfo
+  private case class CommandContext[PS: ScenarioShapeFetchStrategy](
+      processDetails: ScenarioWithDetailsEntity[PS],
```

Comment:
I see a little problem here. `processDetails` can contain scenario shape for the latest version and  `versionOnWhichActionIsDone` is for some other version. Why we can't replace `processRepository.fetchLatestProcessDetailsForProcessId` with `fetchProcessDetailsForId`. Thanks to that we avoid confusion and we can remove `versionOnWhichActionIsDone` as it is already in `ScenarioWithDetailsEntity`.

Reply:
What about that?

Reply:
I like that idea to run action on a scenario in given version (last version to deploy or last deployed version to cancel). There is that one edge-case where "cancel" relies on missing version (scenario -> lastDeployedAction -> processVersionId):
```
            // Currently we don't send notifications and don't add finished action into db if version id is missing.
            // It happens during cancel when there is no finished deploy action before it, but some scenario is running on engine.
            // TODO: We should send notifications and add action db entry in that cases as well as for normal finish.
            //       Before we can do that we should check if we somewhere rely on fact that version is always defined -
            //       see ProcessAction.processVersionId
            logger.info(s"Action $actionString finished for action without version id - skipping listener notification")
            removeInvalidAction(ctx.actionId)
```
WDYT?

Reply:
Ah I forgot about that edge case. Correct me if I'm wrong. Do you propose to not allow to cancel scenario in such situation? What user can do then? Should he/she ask admin to kill such a scenario? Is it possible to make this field somehow optional and handle also this case?

Reply:
2. Also please name it more explicit - `versionOnWhichActionIsDone` part was pretty good, we can adapt it to this situation

Reply:
Please name it latestScenarioDetails to make it clear that it is latest

Reply:
Done, renamed.